File: apps/desktop/src/features/settings/preferences.ts
import { createStore } from "@tanstack/react-store"

import type { SaveStatus } from "@/components/oxyn/appearance-settings"
import type { BackendFailure } from "@/components/oxyn/backend-error-alert"
import { toast } from "@/components/ui/toast"
import { BackendError } from "@/lib/ipc/client"
import { settingsBackend } from "@/lib/ipc/settings"
import type {
  DisplayPreferences,
  PreferencesChange,
  PreferencesState,
} from "@/lib/ipc/settings"

/** Where the last save stands. A failure never reads as saved (UX-SPEC). */
export type SaveState = SaveStatus

export interface PreferencesStore {
  /** Applied locally: what the interface shows, saved or not yet. */
  preferences: DisplayPreferences
  /** False until the backend answered the first read. */
  loaded: boolean
  loadError: BackendFailure | null
  save: SaveState
  /**
   * Counts the saves that changed how the backend formats cells. Result pages
   * are formatted there and cached until invalidated: a new value says the
   * cached ones are formatted the old way.
   */
  formatRevision: number
}

/** The preferences the backend formats cells with (`format_options_of`). */
const FORMAT_FIELDS = [
  "nullText",
  "groupThousands",
  "binaryDisplay",
  "cellMaxChars",
] as const satisfies ReadonlyArray<keyof DisplayPreferences>

/**
 * The domain defaults of `WorkspacePreferences`, shown only until the first
 * read answers. Never written: the backend numbers and merges every write.
 */
export const DEFAULT_PREFERENCES: DisplayPreferences = {
  theme: "dark",
  readingDensity: "compact",
  sidebarCollapsed: false,
  inspectorOpen: true,
  inspectorWidth: 280,
  nullText: "∅ NULL",
  groupThousands: false,
  binaryDisplay: "hex",
  cellMaxChars: 512,
}

export const preferencesStore = createStore<PreferencesStore>({
  preferences: DEFAULT_PREFERENCES,
  loaded: false,
  loadError: null,
  save: { status: "idle" },
  formatRevision: 0,
})

function failureOf(error: unknown): BackendFailure {
  return error instanceof BackendError
    ? { message: error.message, retryable: error.retryable }
    : { message: String(error), retryable: false }
}

let loading: Promise<void> | null = null

/** Reads the preferences once per window load; later calls share the read. */
export function loadPreferences(): Promise<void> {
  loading ??= settingsBackend
    .readPreferences()
    .then((state: PreferencesState) => {
      preferencesStore.setState((current) => ({
        ...current,
        preferences: state.preferences,
        loaded: true,
        loadError: null,
      }))
    })
    .catch((error: unknown) => {
      loading = null
      preferencesStore.setState((current) => ({
        ...current,
        loadError: failureOf(error),
      }))
    })
  return loading
}

// Answers can arrive out of order: only the latest request settles the status.
let latest = 0

/** How one save ended; `superseded` when a later change answers instead. */
export type SaveOutcome =
  | { status: "saved" }
  | { status: "failed"; message: string }
  | { status: "superseded" }

/**
 * Applies a change now, then saves it.
 *
 * Local first — a display setting is local (UX-SPEC « Ce qui n'est jamais
 * optimiste ») — and the save state says whether it reached the workspace.
 */
export async function changePreferences(
  change: PreferencesChange
): Promise<SaveOutcome> {
  const request = ++latest
  preferencesStore.setState((current) => ({
    ...current,
    preferences: { ...current.preferences, ...change },
    save: { status: "saving" },
  }))
  try {
    await settingsBackend.writePreferences(change)
    // Before the superseded check: this write reached the backend whatever
    // answers after it, so pages formatted before it are stale.
    if (FORMAT_FIELDS.some((field) => field in change))
      preferencesStore.setState((current) => ({
        ...current,
        formatRevision: current.formatRevision + 1,
      }))
    if (request !== latest) return { status: "superseded" }
    preferencesStore.setState((current) => ({
      ...current,
      save: { status: "saved" },
    }))
    return { status: "saved" }
  } catch (error) {
    if (request !== latest) return { status: "superseded" }
    const message = failureOf(error).message
    preferencesStore.setState((current) => ({
      ...current,
      save: { status: "failed", message },
    }))
    return { status: "failed", message }
  }
}

/** Saves what is applied again, under a new revision. */
export function retrySavingPreferences() {
  return changePreferences({})
}

/**
 * A change made from the workspace rather than from the settings dialog.
 *
 * The failure is treated as the dialog treats it — applied, said not saved,
 * with `Save again` — but in a toast, since no save status line is on screen
 * there. A failure never reads as saved (ADR-0013).
 */
export async function changePreferencesFromView(change: PreferencesChange) {
  const outcome = await changePreferences(change)
  if (outcome.status !== "failed") return
  toast.add({
    title: "Preference applied here but not saved",
    description: outcome.message,
    type: "warning",
    actionProps: {
      children: "Save again",
      onClick: () => void changePreferencesFromView({}),
    },
  })
}

File: crates/oxyn-desktop/src/backend/settings/preferences.rs
//! Workspace display preferences, saved in revision order
//! ([ADR-0013](../../../../../docs/adr/0013-preferences-workspace.md)).
//!
//! # Why the backend numbers the revisions
//!
//! Tauri runs async commands concurrently: two writes sent 5 ms apart may
//! reach the store in either order. The store keeps the highest revision, so
//! the order that matters is the order in which revisions are **minted** —
//! and minting them here, under one lock, makes that the order of the calls.
//! A front that numbered them would lose that property on its first reload,
//! restarting from a revision it read before a write still in flight.
//!
//! # Why a write survives its caller
//!
//! The dispatch runs in a task of its own and the command awaits it. When the
//! webview reloads or the window closes, the IPC future is dropped but the
//! write goes on, counted as a local write the shutdown waits for
//! ([ADR-0021](../../../../../docs/adr/0021-marqueur-d-arret.md)).

use std::sync::Arc;

use oxyn_core::{
    Actor, BinaryPreference, CancelToken, Command, PreferencesSnapshot, WorkspacePreferences,
};
use oxyn_data::{BinaryDisplay, DEFAULT_MAX_LEN, FormatOptions, NumberGrouping};
use oxyn_exec::Outcome;

use crate::backend::Backend;
use crate::ipc::IpcError;
use crate::ipc::settings::{
    DisplayPreferences, PreferencesChange, PreferencesSaved, PreferencesState,
};

#[derive(Default)]
pub(crate) struct PreferenceState {
    /// The latest snapshot applied locally, saved or not yet.
    applied: parking_lot::Mutex<PreferencesSnapshot>,
    /// Whether `applied` came from the store. An async lock, held across the
    /// read: a write that started from defaults would store them over the
    /// user's choices.
    loaded: tokio::sync::Mutex<bool>,
}

fn state_of(snapshot: &PreferencesSnapshot) -> PreferencesState {
    PreferencesState {
        revision: snapshot.revision,
        preferences: DisplayPreferences::of(&snapshot.preferences),
    }
}

/// The grid's cell rendering for a set of preferences.
///
/// The one place a preference becomes [`FormatOptions`]: the front never
/// reformats a cell, so this is what the user sees.
pub(crate) fn format_options_of(preferences: &WorkspacePreferences) -> FormatOptions {
    FormatOptions::default()
        .with_null_text(preferences.null_text.clone())
        .with_number_grouping(if preferences.group_thousands {
            NumberGrouping::Thousands
        } else {
            NumberGrouping::None
        })
        .with_binary_display(match preferences.binary_display {
            BinaryPreference::Base64 => BinaryDisplay::Base64,
            BinaryPreference::Size => BinaryDisplay::Size,
            _ => BinaryDisplay::Hex,
        })
        .with_max_len(usize::try_from(preferences.cell_max_chars).unwrap_or(DEFAULT_MAX_LEN))
}

impl Backend {
    /// The preferences, read from the store on the first call.
    ///
    /// # Errors
    /// If the stored payload is unreadable. It is then never replaced: every
    /// write fails the same way until the state is repaired.
    pub async fn read_preferences(&self) -> Result<PreferencesState, IpcError> {
        self.ensure_preferences_loaded().await?;
        Ok(state_of(&self.inner.settings.preferences.applied.lock()))
    }

    /// Applies a change and saves it under the next revision.
    ///
    /// The change is applied locally before the save answers: a display
    /// setting is local ([UX-SPEC](../../../../../docs/UX-SPEC.md#ce-qui-nest-jamais-optimiste)).
    /// A failed save leaves it applied and says so; saving again is an empty
    /// change.
    ///
    /// # Errors
    /// An invalid value (refused before anything is applied), a store failure,
    /// or a newer revision written elsewhere.
    pub async fn write_preferences(
        &self,
        change: PreferencesChange,
    ) -> Result<PreferencesSaved, IpcError> {
        self.save_preferences(|preferences| change.apply(preferences))
            .await
    }

    /// Applies `change` to what is applied now and saves it under the next
    /// revision: the one path every preference write takes, the display
    /// settings' and the restored object's alike.
    pub(super) async fn save_preferences(
        &self,
        change: impl FnOnce(&mut WorkspacePreferences),
    ) -> Result<PreferencesSaved, IpcError> {
        self.ensure_preferences_loaded().await?;
        let state = &self.inner.settings.preferences;
        let snapshot = {
            let mut applied = state.applied.lock();
            let mut next = applied.clone();
            change(&mut next.preferences);
            next.revision = applied
                .revision
                .checked_add(1)
                .filter(|revision| i64::try_from(*revision).is_ok())
                .ok_or_else(|| {
                    IpcError::invalid("The preference revision is exhausted; reload preferences")
                })?;
            next.validate()?;
            applied.clone_from(&next);
            next
        };
        let revision = snapshot.revision;

        let command = Command::WriteWorkspacePreferences {
            workspace: self.inner.executor.workspace(),
            snapshot: Box::new(snapshot),
        };
        let guard = self.local_write(&command);
        let inner = Arc::clone(&self.inner);
        let task = tauri::async_runtime::spawn(async move {
            let _guard = guard;
            inner
                .executor
                .dispatch(Actor::Human, command, &CancelToken::new())
                .await
        });
        let outcome = task
            .await
            .map_err(|_| IpcError::invalid("The preference writer stopped"))??;

        match outcome {
            Outcome::WorkspacePreferences { snapshot: saved } => {
                let mut applied = state.applied.lock();
                if saved.revision > applied.revision {
                    // Neither this write nor a later local one: another
                    // instance saved. The next local write starts above it.
                    applied.revision = saved.revision;
                    return Err(IpcError::invalid(
                        "Preferences changed elsewhere. Save again to apply your current choices.",
                    ));
                }
                Ok(PreferencesSaved {
                    revision,
                    saved: state_of(&saved),
                })
            }
            Outcome::Denied { reason, .. } => Err(IpcError::invalid(reason)),
            _ => Err(IpcError::invalid("Unexpected preference save response")),
        }
    }

    /// How cells are formatted now, for every page the grid reads.
    ///
    /// Defaults until the preferences were read: the front reads them at
    /// startup, before a result can exist.
    #[must_use]
    pub fn format_options(&self) -> FormatOptions {
        format_options_of(&self.inner.settings.preferences.applied.lock().preferences)
    }

    async fn ensure_preferences_loaded(&self) -> Result<(), IpcError> {
        let state = &self.inner.settings.preferences;
        let mut loaded = state.loaded.lock().await;
        if *loaded {
            return Ok(());
        }
        let executor = &self.inner.executor;
        let outcome = executor
            .dispatch(
                Actor::Human,
                Command::ReadWorkspacePreferences {
                    workspace: executor.workspace(),
                },
                &CancelToken::new(),
            )
            .await?;
        let Outcome::WorkspacePreferences { snapshot } = outcome else {
            return Err(IpcError::invalid("Unexpected preference read response"));
        };
        *state.applied.lock() = snapshot;
        *loaded = true;
        Ok(())
    }
}
